File: src/main/java/com/atsuishio/superbwarfare/init/ModMenuTypes.java
package com.atsuishio.superbwarfare.init;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.menu.*;
import net.minecraft.world.inventory.MenuType;

import java.util.function.Supplier;

public class ModMenuTypes {

    public static final Supplier<MenuType<ReforgingTableMenu>> REFORGING_TABLE_MENU =
            Registration.menu("reforging_table_menu",
                    () -> new MenuType<>((windowId, inv) -> new ReforgingTableMenu(windowId, inv)));
    public static final Supplier<MenuType<ChargingStationMenu>> CHARGING_STATION_MENU =
            Registration.menu("charging_station_menu",
                    () -> new MenuType<>((windowId, inv) -> new ChargingStationMenu(windowId, inv)));

    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_MINI =
            Registration.menu("vehicle_menu_mini",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.mini(windowId, inv, false)));
    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_MINI_UPGRADE =
            Registration.menu("vehicle_menu_mini_upgrade",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.mini(windowId, inv, true)));

    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_SMALL =
            Registration.menu("vehicle_menu_small",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.small(windowId, inv, false)));
    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_SMALL_UPGRADE =
            Registration.menu("vehicle_menu_small_upgrade",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.small(windowId, inv, true)));

    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_MEDIUM =
            Registration.menu("vehicle_menu_medium",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.medium(windowId, inv, false)));
    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_MEDIUM_UPGRADE =
            Registration.menu("vehicle_menu_medium_upgrade",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.medium(windowId, inv, true)));

    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_LARGE =
            Registration.menu("vehicle_menu_large",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.large(windowId, inv, false)));
    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_LARGE_UPGRADE =
            Registration.menu("vehicle_menu_large_upgrade",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.large(windowId, inv, true)));

    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_HUGE =
            Registration.menu("vehicle_menu_huge",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.huge(windowId, inv, false)));
    public static final Supplier<MenuType<VehicleMenu>> VEHICLE_MENU_HUGE_UPGRADE =
            Registration.menu("vehicle_menu_huge_upgrade",
                    () -> new MenuType<>((windowId, inv) -> VehicleMenu.huge(windowId, inv, true)));

    public static final Supplier<MenuType<SuperbItemInterfaceMenu>> SUPERB_ITEM_INTERFACE_MENU =
            Registration.menu("superb_item_interface_menu",
                    () -> new MenuType<>((windowId, inv) -> new SuperbItemInterfaceMenu(windowId, inv)));
    public static final Supplier<MenuType<FuMO25Menu>> FUMO_25_MENU =
            Registration.menu("fumo_25_menu",
                    () -> new MenuType<>((windowId, inv) -> new FuMO25Menu(windowId, inv)));
    public static final Supplier<MenuType<VehicleAssemblingMenu>> VEHICLE_ASSEMBLING_MENU =
            Registration.menu("vehicle_assembling_menu",
                    () -> new MenuType<>((windowId, inv) -> new VehicleAssemblingMenu(windowId, inv)));

    public static void register() {

    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/init/ModMenuTypes.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.inventory.menu.BlueprintResearchTableMenu
import com.atsuishio.superbwarfare.inventory.menu.ChargingStationMenu
import com.atsuishio.superbwarfare.inventory.menu.FuMO25Menu
import com.atsuishio.superbwarfare.inventory.menu.HugeVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.LargeVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.MediumVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.MiniVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.ReforgingTableMenu
import com.atsuishio.superbwarfare.inventory.menu.SmallVehicleContainerMenu
import com.atsuishio.superbwarfare.inventory.menu.SuperbItemInterfaceMenu
import com.atsuishio.superbwarfare.inventory.menu.VehicleAssemblingMenu
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.flag.FeatureFlagSet
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.MenuType

@Suppress("unused")
object ModMenuTypes {
    @JvmField
    val REFORGING_TABLE_MENU: MenuType<ReforgingTableMenu> =
        register("reforging_table_menu", MenuType.MenuSupplier { windowId, inv ->
            ReforgingTableMenu(windowId, inv)
        })

    @JvmField
    val CHARGING_STATION_MENU: MenuType<ChargingStationMenu> =
        register("charging_station_menu", MenuType.MenuSupplier { windowId, inv ->
            ChargingStationMenu(windowId, inv)
        })

    @JvmField
    val MINI_VEHICLE_CONTAINER_MENU: MenuType<MiniVehicleContainerMenu> =
        register("mini_vehicle_container", MiniVehicleContainerMenu.TYPE)

    @JvmField
    val SMALL_VEHICLE_CONTAINER_MENU: MenuType<SmallVehicleContainerMenu> =
        register("small_vehicle_container", SmallVehicleContainerMenu.TYPE)

    @JvmField
    val MEDIUM_VEHICLE_CONTAINER_MENU: MenuType<MediumVehicleContainerMenu> =
        register("medium_vehicle_container", MediumVehicleContainerMenu.TYPE)

    @JvmField
    val LARGE_VEHICLE_CONTAINER_MENU: MenuType<LargeVehicleContainerMenu> =
        register("large_vehicle_container", LargeVehicleContainerMenu.TYPE)

    @JvmField
    val HUGE_VEHICLE_CONTAINER_MENU: MenuType<HugeVehicleContainerMenu> =
        register("huge_vehicle_container", HugeVehicleContainerMenu.TYPE)

    @JvmField
    val SUPERB_ITEM_INTERFACE_MENU: MenuType<SuperbItemInterfaceMenu> =
        register("superb_item_interface_menu", MenuType.MenuSupplier { windowId, inv ->
            SuperbItemInterfaceMenu(windowId, inv)
        })

    @JvmField
    val FUMO_25_MENU: MenuType<FuMO25Menu> =
        register("fumo_25_menu", MenuType.MenuSupplier { windowId, inv ->
            FuMO25Menu(windowId, inv)
        })

    @JvmField
    val VEHICLE_ASSEMBLING_MENU: MenuType<VehicleAssemblingMenu> =
        register("vehicle_assembling_menu", MenuType.MenuSupplier { windowId, inv ->
            VehicleAssemblingMenu(windowId, inv)
        })

    @JvmField
    val BLUEPRINT_RESEARCH_TABLE: MenuType<BlueprintResearchTableMenu> =
        register("blueprint_research_table_menu", MenuType.MenuSupplier { windowId, inv ->
            BlueprintResearchTableMenu(windowId, inv)
        })

    private fun <T : AbstractContainerMenu> register(
        name: String,
        factory: MenuType.MenuSupplier<T>
    ): MenuType<T> {
        return register(name, MenuType(factory, FeatureFlagSet.of()))
    }

    private fun <T : AbstractContainerMenu> register(
        name: String,
        type: MenuType<T>
    ): MenuType<T> {
        return Registry.register(BuiltInRegistries.MENU, Mod.loc(name), type)
    }

    @JvmStatic
    fun init() {
    }
}
